File: packages/@aws-cdk/logs/lib/cross-account-destination.ts
import cdk = require('@aws-cdk/core');
import iam = require('@aws-cdk/iam');
import { LogGroup } from './log-group';
import { cloudformation, DestinationArn } from './logs.generated';
import { ISubscriptionDestination, SubscriptionDestination } from './subscription-filter';

export interface DestinationProps {
    /**
     * The name of the log destination.
     */
    destinationName: string;

    /**
     * The role to assume that grants permissions to write to 'target'.
     *
     * The role must be assumable by 'logs.{REGION}.amazonaws.com'.
     */
    role: iam.Role;

    /**
     * The log destination target's ARN
     */
    targetArn: cdk.Arn;
}

/**
 * Create a new CloudWatch Logs Destination.
 *
 * Log destinations can be used to subscribe a Kinesis stream in a different
 * account to a CloudWatch Subscription. A Kinesis stream in the same account
 * can be subscribed directly.
 *
 * The @aws-cdk/kinesis library takes care of this automatically; you shouldn't
 * need to bother with this class.
 */
export class CrossAccountDestination extends cdk.Construct implements ISubscriptionDestination {
    public readonly policyDocument: cdk.PolicyDocument = new cdk.PolicyDocument();
    public readonly destinationName: DestinationName;
    public readonly destinationArn: DestinationArn;

    constructor(parent: cdk.Construct, id: string, props: DestinationProps) {
        super(parent, id);

        this.policyDocument = new cdk.PolicyDocument();

        const resource = new cloudformation.DestinationResource(this, 'Resource', {
            destinationName: props.destinationName,
            destinationPolicy: new cdk.Token(() => !this.policyDocument.isEmpty ? JSON.stringify(this.policyDocument.resolve()) : ""),
            roleArn: props.role.roleArn,
            targetArn: props.targetArn
        });

        this.destinationArn = resource.destinationArn;
        this.destinationName = resource.ref;
    }

    public addToPolicy(statement: cdk.PolicyStatement) {
        this.policyDocument.addStatement(statement);
    }

    public subscriptionDestination(_sourceLogGroup: LogGroup): SubscriptionDestination {
        return { arn: this.destinationArn };
    }
}

/**
 * Name of a CloudWatch Destination
 */
export class DestinationName extends cdk.Token {
}

File: packages/aws-cdk-dotnet/src/Amazon.CDK.AWS.Logs/CrossAccountDestination.cs
using Amazon.CDK;
using AWS.Jsii.Runtime.Deputy;

namespace Amazon.CDK.AWS.Logs
{
    /// <summary>
    /// A new CloudWatch Logs Destination for use in cross-account scenarios
    /// 
    /// Log destinations can be used to subscribe a Kinesis stream in a different
    /// account to a CloudWatch Subscription. A Kinesis stream in the same account
    /// can be subscribed directly.
    /// 
    /// The @aws-cdk/aws-kinesis library takes care of this automatically; you shouldn't
    /// need to bother with this class.
    /// </summary>
    [JsiiClass(typeof(CrossAccountDestination), "@aws-cdk/aws-logs.CrossAccountDestination", "[{\"name\":\"parent\",\"type\":{\"fqn\":\"@aws-cdk/cdk.Construct\"}},{\"name\":\"id\",\"type\":{\"primitive\":\"string\"}},{\"name\":\"props\",\"type\":{\"fqn\":\"@aws-cdk/aws-logs.CrossAccountDestinationProps\"}}]")]
    public class CrossAccountDestination : Construct, IILogSubscriptionDestination
    {
        public CrossAccountDestination(Construct parent, string id, ICrossAccountDestinationProps props): base(new DeputyProps(new object[]{parent, id, props}))
        {
        }

        protected CrossAccountDestination(ByRefValue reference): base(reference)
        {
        }

        protected CrossAccountDestination(DeputyProps props): base(props)
        {
        }

        /// <summary>Policy object of this CrossAccountDestination object</summary>
        [JsiiProperty("policyDocument", "{\"fqn\":\"@aws-cdk/cdk.PolicyDocument\"}")]
        public virtual PolicyDocument PolicyDocument
        {
            get => GetInstanceProperty<PolicyDocument>();
        }

        /// <summary>The name of this CrossAccountDestination object</summary>
        [JsiiProperty("destinationName", "{\"fqn\":\"@aws-cdk/aws-logs.DestinationName\"}")]
        public virtual DestinationName DestinationName
        {
            get => GetInstanceProperty<DestinationName>();
        }

        /// <summary>The ARN of this CrossAccountDestination object</summary>
        [JsiiProperty("destinationArn", "{\"fqn\":\"@aws-cdk/aws-logs.DestinationArn\"}")]
        public virtual DestinationArn DestinationArn
        {
            get => GetInstanceProperty<DestinationArn>();
        }

        [JsiiMethod("addToPolicy", null, "[{\"name\":\"statement\",\"type\":{\"fqn\":\"@aws-cdk/cdk.PolicyStatement\"}}]")]
        public virtual void AddToPolicy(PolicyStatement statement)
        {
            InvokeInstanceVoidMethod(new object[]{statement});
        }

        /// <summary>
        /// Return the properties required to send subscription events to this destination.
        /// 
        /// If necessary, the destination can use the properties of the SubscriptionFilter
        /// object itself to configure its permissions to allow the subscription to write
        /// to it.
        /// 
        /// The destination may reconfigure its own permissions in response to this
        /// function call.
        /// </summary>
        [JsiiMethod("logSubscriptionDestination", "{\"fqn\":\"@aws-cdk/aws-logs.LogSubscriptionDestination\"}", "[{\"name\":\"_sourceLogGroup\",\"type\":{\"fqn\":\"@aws-cdk/aws-logs.LogGroup\"}}]")]
        public virtual ILogSubscriptionDestination LogSubscriptionDestination(LogGroup _sourceLogGroup)
        {
            return InvokeInstanceMethod<ILogSubscriptionDestination>(new object[]{_sourceLogGroup});
        }
    }
}